[Core] Support GPU provisioning with flex-start in GKE (#6125)
skypilot-org/skypilot: docs/source/reservations/reservations.rst

.. _dws-on-gke:

Using DWS on GKE with Kueue
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Using DWS on GKE
~~~~~~~~~~~~~~~~~

DWS is also supported on Google Kubernetes Engine (GKE) with Kueue. To enable DWS on GKE, you need to set up your GKE cluster with Kueue and DWS; see the `GCP official docs <https://cloud.google.com/kubernetes-engine/docs/how-to/provisioningrequest>`__.
GKE supports the two kinds of flex-start configurations:

> To launch a SkyPilot cluster or job on GKE with DWS, you can specify the DWS configuration in the SkyPilot task YAML:

Author: Remove the example usage here so we do not need to maintain duplicate contents in different files.

Reviewer: Better to keep it to make doc self-contained. People don't like jumping around docs.

Author: Updated

> * Flex-start with queued provisioning, where GKE allocates all requested resources at the same time. To use this configuration, you have to add the `--flex-start` and `enable-queued-provisioning` flags when you create the node pool. 

Reviewer: mention some examples for when each mode is best suited. Also, considering SkyPilot clusters are always gang-scheduled, it does not make sense to use `flex-start` with `kueue` setup, right?

Author: Updated

> Refer to the `GKE DWS example <https://github.com/skypilot-org/skypilot/blob/master/examples/gke_dws/README.md>`_ for the detailed usage of DWS on GKE with SkyPilot.

Reviewer: can we inline the example here to avoid too many docs scattered around?

.. _dws-on-gke:

Using DWS on GKE
~~~~~~~~~~~~~~~~~

Flex-start, powered by `Dynamic Workload Scheduler <https://cloud.google.com/blog/products/compute/introducing-dynamic-workload-scheduler>`_, provides a flexible and cost-effective technique to obtain GPUs when you need to run AI/ML workloads.

> GKE supports the two kinds of flex-start configurations:

Reviewer: Considering we don't expose the `mode` any more. Should we update the doc? We should not have the detailed explanation for flex-start here, considering we will automatically select. We can have the explanation in a dropdown

Author: Updated

  GKE supports the two kinds of flex-start configurations:

  * Flex-start, where GKE allocates resources node by node.

    * This configuration only requires you to set the ``flex-start`` flag during node creation. 

>     * Flex-start configuration is recommended for the small to medium workload size, which means that the workload can run on a single node. For example, this configuration works well if you are running small training jobs, offline inference, or batch jobs.

Reviewer 2: ```suggestion
    * Flex-start is recommended for small to medium (single-node) workloads such as small training jobs, offline inference, or batch jobs.
```

Author: Updated

  * Flex-start with queued provisioning, where GKE allocates all requested resources at the same time.

>     * To use this configuration, you have to add the ``flex-start`` and ``enable-queued-provisioning`` flags when you create the node pool.

Reviewer 2: ```suggestion
    * To use this configuration, add the ``--flex-start`` and ``--enable-queued-provisioning`` flags when you create the node pool.
```

>     * Flex-start with queued provisioning configuration is recommended for the medium to large workload, which means that the workload can run on multiple nodes. Your workload requires multiple resources and can't start running until all nodes are provisioned and ready at the same time. For example, this configuration works well if you are running distributed machine learning training workloads.

Reviewer 2: ```suggestion
    * Flex-start with queued provisioning is recommended for multi-node workloads, which require multiple resources and can't start running until all nodes are provisioned and ready at the same time. For example, distributed machine learning training workloads.
```

Author: Updated

  See `GKE DWS documentation <https://cloud.google.com/kubernetes-engine/docs/concepts/dws>`_ for more details.

  .. note::

>       If you are using Kueue, and ``kubernetes.kueue.local_queue_name`` is specified, SkyPilot will automatically enable the flex-start with queued provisioning mode, otherwise it will use the flex-start mode.

Author: Updated

2. Configure the following fields in ``~/.sky/config.yaml``:

.. code-block:: yaml

    kubernetes:
      # provision_timeout: 1200
      autoscaler: gke
      dws:
        enabled: true

When flex-start is enabled, the default ``provision_timeout`` is set to ``600`` seconds (10 minutes). If you encounter provisioning timeout issues, you can increase this value in your configuration. For example, set it to ``1200`` seconds (20 minutes) or higher depending on your workload and cluster size.

3. Launch your clusters or managed jobs.

.. code-block:: yaml

    name: flex-start

    resources:
      infra: k8s
      accelerators: L4:1

    num_nodes: 1

In case you want to specify the DWS configuration for each job/cluster, you can also specify the configuration in the SkyPilot task YAML (see :ref:`here <config-client-job-task-yaml>`):

.. code-block:: yaml

    name: flex-start

    resources:
      infra: k8s
      accelerators: L4:1

    num_nodes: 1

    config:
      kubernetes:
        dws:
          enabled: true

> Flex-start with queued provisioning

Reviewer: The docs should be changed, i.e. we should not separate the case by `Flex-start with queued provisioning` but using `Kueue` or not.
